File: source/ts/src/astronomy/riseSet.ts
/** Position is fitted in **rectangular** coordinates, never RA: a fit across the 24 h seam is wrong for days at a time. */
import { sin, cos } from './trig';
import { getMoonPositionForTrack } from './moon';
import { getSunPosition } from './sun';
import { ttDaysSinceJ2000 } from './deltaT';
import { nutation, meanObliquityArcsec, ARCSEC_TO_RAD, AU_KM } from './frame';
import {
  REFRACTION_NEAR_HORIZON_DEG, SUN_RADIUS_AU, MOON_RADIUS_KM,
  EARTH_EQUATORIAL_RADIUS_KM, EARTH_FLATTENING_SQUARED,
} from './topocentric';
// Cyclic with `riseSetCache.ts`; safe only as a hoisted function declaration.
import { clearRiseSetEventCache } from './riseSetCache';
import type { GeoLocation } from '../types/location';

const DAY_MS = 86_400_000;
const DEG_TO_RAD = Math.PI / 180;
const RAD_TO_DEG = 180 / Math.PI;
const MOON_RADIUS_AU = MOON_RADIUS_KM / AU_KM;

export type RiseSetBody = 'sun' | 'moon';

/** Do not widen: ELP carries argument families near a 5-day period, and 8 days is a cliff more nodes do not rescue. */
const TRACK_BLOCK_DAYS = 4;
const TRACK_NODES = 11;

/** Narrow events are caught by the slope bound, not by this step. */
const SCAN_STEP_DAYS = 12 / (60 * 24);

/** Bound on |d(altitude)/dt|: Earth's 360°/day plus ~13° for the fastest body. */
const MAX_ALTITUDE_SLOPE_DEG_PER_DAY = 380;

const ROOT_TOLERANCE_DAYS = 1e-8;

class PositionTrack {
  private readonly nodeX = new Float64Array(TRACK_NODES);
  private readonly weight = new Float64Array(TRACK_NODES);
  /** Equatorial-of-date x, y, z (AU) at each node. */
  private readonly px = new Float64Array(TRACK_NODES);
  private readonly py = new Float64Array(TRACK_NODES);
  private readonly pz = new Float64Array(TRACK_NODES);
  private readonly midMs: number;
  private readonly halfMs: number;
  readonly radiusAu: number;

  constructor(body: RiseSetBody, blockIndex: number) {
    const startMs = blockIndex * TRACK_BLOCK_DAYS * DAY_MS;
    const endMs = startMs + TRACK_BLOCK_DAYS * DAY_MS;
    this.midMs = (startMs + endMs) / 2;
    this.halfMs = (endMs - startMs) / 2;
    this.radiusAu = body === 'sun' ? SUN_RADIUS_AU : MOON_RADIUS_AU;

    for (let k = 0; k < TRACK_NODES; k++) {
      const x = Math.cos((Math.PI * k) / (TRACK_NODES - 1));
      this.nodeX[k] = x;
      this.weight[k] = (k === 0 || k === TRACK_NODES - 1 ? 0.5 : 1) * (k % 2 ? -1 : 1);

      const ms = this.midMs + this.halfMs * x;
      const date = new Date(ms);
      const t = ttDaysSinceJ2000(date) / 36525;
      const eps = (meanObliquityArcsec(t) + nutation(t).deps) * ARCSEC_TO_RAD;

      let lonDeg: number, latDeg: number, distAu: number;
      if (body === 'sun') {
        const p = getSunPosition(date);
        lonDeg = p.longitude; latDeg = p.latitude; distAu = p.distance;
      } else {
        const p = getMoonPositionForTrack(date);
        lonDeg = p.longitude; latDeg = p.latitude; distAu = p.distance / AU_KM;
      }
      const lon = lonDeg * DEG_TO_RAD;
      const lat = latDeg * DEG_TO_RAD;
      const cosLat = distAu * Math.cos(lat);
      const ex = cosLat * Math.cos(lon);
      const ey = cosLat * Math.sin(lon);
      const ez = distAu * Math.sin(lat);
      this.px[k] = ex;
      this.py[k] = Math.cos(eps) * ey - Math.sin(eps) * ez;
      this.pz[k] = Math.sin(eps) * ey + Math.cos(eps) * ez;
    }
  }

  position(ms: number, out: Float64Array): void {
    const x = (ms - this.midMs) / this.halfMs;
    let nx = 0, ny = 0, nz = 0, den = 0;
    for (let k = 0; k < TRACK_NODES; k++) {
      const dx = x - (this.nodeX[k] as number);
      if (dx === 0) {
        out[0] = this.px[k] as number;
        out[1] = this.py[k] as number;
        out[2] = this.pz[k] as number;
        return;
      }
      const q = (this.weight[k] as number) / dx;
      nx += q * (this.px[k] as number);
      ny += q * (this.py[k] as number);
      nz += q * (this.pz[k] as number);
      den += q;
    }
    out[0] = nx / den;
    out[1] = ny / den;
    out[2] = nz / den;
  }
}

/** Per-day, not per-block, so the two days meeting at a midnight agree exactly there. */
class DayFrame {
  /** Arcseconds, at the day's two endpoints. */
  private readonly eqeq0: number;
  private readonly eqeq1: number;
  /** ΔT in days. Per day, not shared scratch: that makes sidereal time build-order dependent. */
  private readonly deltaTDays: number;
  private readonly dayStartMs: number;

  constructor(dayIndex: number) {
    this.dayStartMs = dayIndex * DAY_MS;
    const dayEndMs = this.dayStartMs + DAY_MS;
    const midMs = (this.dayStartMs + dayEndMs) / 2;
    this.eqeq0 = equationOfEquinoxes(this.dayStartMs);
    this.eqeq1 = equationOfEquinoxes(dayEndMs);
    this.deltaTDays = ttDaysSinceJ2000(new Date(midMs))
      - (midMs - Date.UTC(2000, 0, 1, 12)) / DAY_MS;
  }

  /** Greenwich Apparent Sidereal Time, degrees. */
  gast(ms: number): number {
    const utDays = (ms - Date.UTC(2000, 0, 1, 12)) / DAY_MS;
    const theta = 360 * (((0.7790572732640 + 0.00273781191135448 * utDays) % 1 + (utDays % 1)) % 1);
    const frac = (ms - this.dayStartMs) / DAY_MS;
    const eqeq = this.eqeq0 + (this.eqeq1 - this.eqeq0) * frac;
    // The accumulated-precession polynomial takes TT, never UT.
    const t = (utDays + this.deltaTDays) / 36525;
    const precession = 0.014506
      + (4612.156534 + (1.3915817 + (-0.00000044 + (-0.000029956 + -0.0000000368 * t) * t) * t) * t) * t;
    const gast = (theta + (eqeq + precession) / 3600) % 360;
    return gast < 0 ? gast + 360 : gast;
  }
}

function equationOfEquinoxes(ms: number): number {
  const t = ttDaysSinceJ2000(new Date(ms)) / 36525;
  const { dpsi, deps } = nutation(t);
  return dpsi * Math.cos((meanObliquityArcsec(t) + deps) * ARCSEC_TO_RAD);
}

const TRACK_STORE = new Map<string, PositionTrack>();
const MAX_TRACKS = 1024;
const FRAME_STORE = new Map<number, DayFrame>();
const MAX_FRAMES = 8192;

function trackFor(body: RiseSetBody, dayIndex: number): PositionTrack {
  const blockIndex = Math.floor(dayIndex / TRACK_BLOCK_DAYS);
  const key = `${body}|${blockIndex}`;
  const existing = TRACK_STORE.get(key);
  if (existing !== undefined) return existing;
  const track = new PositionTrack(body, blockIndex);
  if (TRACK_STORE.size >= MAX_TRACKS) TRACK_STORE.clear();
  TRACK_STORE.set(key, track);
  return track;
}

function frameFor(dayIndex: number): DayFrame {
  const existing = FRAME_STORE.get(dayIndex);
  if (existing !== undefined) return existing;
  const frame = new DayFrame(dayIndex);
  if (FRAME_STORE.size >= MAX_FRAMES) FRAME_STORE.clear();
  FRAME_STORE.set(dayIndex, frame);
  return frame;
}

/** The fourth store lives in `riseSetCache.ts` and must stay in this list. */
export function clearRiseSetTracks(): void {
  TRACK_STORE.clear();
  FRAME_STORE.clear();
  SCAN_CACHE.clear();
  clearRiseSetEventCache();
}

const BODY_VEC = new Float64Array(3);

class ObserverGeometry {
  readonly sinPhi: number;
  readonly cosPhi: number;
  /** Distance from the Earth's axis, and from its equatorial plane, in AU. */
  readonly equatorialAu: number;
  readonly polarAu: number;
  readonly longitude: number;

  constructor(location: GeoLocation) {
    const phi = location.latitude * DEG_TO_RAD;
    this.sinPhi = Math.sin(phi);
    this.cosPhi = Math.cos(phi);
    const flattened = EARTH_FLATTENING_SQUARED * this.sinPhi * this.sinPhi;
    const c = 1 / Math.sqrt(this.cosPhi * this.cosPhi + flattened);
    const s = EARTH_FLATTENING_SQUARED * c;
    const heightKm = (location.elevation ?? 0) / 1000;
    this.equatorialAu = (EARTH_EQUATORIAL_RADIUS_KM * c + heightKm) / AU_KM * this.cosPhi;
    this.polarAu = (EARTH_EQUATORIAL_RADIUS_KM * s + heightKm) / AU_KM * this.sinPhi;
    this.longitude = location.longitude;
  }
}

/** The **upper limb** above the refracted horizon. Parallax is not a separate term: it falls out of subtracting the observer's position. */
function altitudeExcess(
  track: PositionTrack, frame: DayFrame, geometry: ObserverGeometry, ms: number,
): number {
  track.position(ms, BODY_VEC);
  const local = (frame.gast(ms) + geometry.longitude) * DEG_TO_RAD;
  // `trig.ts`, not `Math.cos`: innermost expression in the module.
  const cosLocal = cos(local);
  const sinLocal = sin(local);

  const x = (BODY_VEC[0] as number) - geometry.equatorialAu * cosLocal;
  const y = (BODY_VEC[1] as number) - geometry.equatorialAu * sinLocal;
  const z = (BODY_VEC[2] as number) - geometry.polarAu;
  // Not `Math.hypot`: V8 will not inline it, and nothing here can overflow.
  const distance = Math.sqrt(x * x + y * y + z * z);

  const dot = (x * geometry.cosPhi * cosLocal
    + y * geometry.cosPhi * sinLocal
    + z * geometry.sinPhi) / distance;
  const centre = Math.asin(dot < -1 ? -1 : dot > 1 ? 1 : dot) * RAD_TO_DEG;
  return centre + (track.radiusAu / distance) * RAD_TO_DEG + REFRACTION_NEAR_HORIZON_DEG;
}

function refine(
  f: (ms: number) => number, loMs: number, hiMs: number, fLoIn: number, wantRise: boolean,
): number {
  const tolMs = ROOT_TOLERANCE_DAYS * DAY_MS;
  let lo = loMs, hi = hiMs, fLo = fLoIn;
  const below = (v: number): boolean => (wantRise ? v < 0 : v >= 0);
  while (hi - lo > tolMs) {
    const mid = (lo + hi) / 2;
    const fMid = f(mid);
    if (below(fLo) === below(fMid)) { lo = mid; fLo = fMid; } else hi = mid;
  }
  return (lo + hi) / 2;
}

interface DayEventPair {
  rise: number[];
  set: number[];
}

/** At high latitude the Moon can rise and set again inside minutes, so a cell is *proved* empty by the slope bound, never assumed. */
function scanDay(body: RiseSetBody, location: GeoLocation, dayIndex: number): DayEventPair {
  const track = trackFor(body, dayIndex);
  const frame = frameFor(dayIndex);
  const geometry = new ObserverGeometry(location);
  const dayStart = dayIndex * DAY_MS;
  const dayEnd = dayStart + DAY_MS;
  const f = (ms: number): number => altitudeExcess(track, frame, geometry, ms);

  const rise: number[] = [];
  const set: number[] = [];

  const scan = (loMs: number, hiMs: number, fLo: number, fHi: number, depth: number): void => {
    if (fLo < 0 && fHi >= 0) { rise.push(refine(f, loMs, hiMs, fLo, true)); return; }
    if (fLo >= 0 && fHi < 0) { set.push(refine(f, loMs, hiMs, fLo, false)); return; }
    if (depth >= 12) return;

    const span = MAX_ALTITUDE_SLOPE_DEG_PER_DAY * ((hiMs - loMs) / DAY_MS);
    const highest = (fLo + fHi + span) / 2;
    const lowest = (fLo + fHi - span) / 2;
    if (fLo < 0 ? highest < 0 : lowest >= 0) return;

    const midMs = (loMs + hiMs) / 2;
    const fMid = f(midMs);
    scan(loMs, midMs, fLo, fMid, depth + 1);
    scan(midMs, hiMs, fMid, fHi, depth + 1);
  };

  const stepMs = SCAN_STEP_DAYS * DAY_MS;
  let prevMs = dayStart;
  let prevF = f(prevMs);
  for (let ms = dayStart + stepMs; ; ms += stepMs) {
    const capped = Math.min(ms, dayEnd);
    const value = f(capped);
    scan(prevMs, capped, prevF, value, 0);
    prevMs = capped;
    prevF = value;
    if (capped >= dayEnd) break;
  }

  // A root refined at the edge can land a hair outside the UTC day it belongs to.
  const inside = (list: number[]): number[] =>
    list.filter((e) => e >= dayStart && e < dayEnd).sort((a, b) => a - b);
  return { rise: inside(rise), set: inside(set) };
}

const SCAN_CACHE = new Map<string, DayEventPair>();
const MAX_SCANS = 20_000;

/** Every rise (`+1`) or set (`-1`) of `body` inside UTC day `dayIndex`, ascending. */
export function dayEvents(
  body: RiseSetBody, direction: 1 | -1, location: GeoLocation, dayIndex: number,
): readonly number[] {
  const key = `${body}|${location.latitude}|${location.longitude}|${location.elevation ?? 0}|${dayIndex}`;
  let pair = SCAN_CACHE.get(key);
  if (pair === undefined) {
    pair = scanDay(body, location, dayIndex);
    if (SCAN_CACHE.size >= MAX_SCANS) SCAN_CACHE.clear();
    SCAN_CACHE.set(key, pair);
  }
  return direction === 1 ? pair.rise : pair.set;
}

File: source/go/internal/astronomy/riseset.go
package astronomy

import (
	"math"
	"sort"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/store"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
)

const moonRadiusAU = MoonRadiusKm / AuKm

type RiseSetBody string

const (
	RiseSetSun  RiseSetBody = "sun"
	RiseSetMoon RiseSetBody = "moon"
)

const (
	trackBlockDays = 4
	trackNodes     = 11
)

const scanStepDays float64 = 12.0 / (60 * 24)

const maxAltitudeSlopeDegPerDay = 380

const rootToleranceDays float64 = 1e-8

type positionTrack struct {
	nodeX    [trackNodes]float64
	weight   [trackNodes]float64
	px       [trackNodes]float64
	py       [trackNodes]float64
	pz       [trackNodes]float64
	midMs    float64
	halfMs   float64
	radiusAu float64
}

func newPositionTrack(ctx *EphemerisCtx, body RiseSetBody, blockIndex int64) *positionTrack {
	startMs := float64(blockIndex * trackBlockDays * dayMS)
	endMs := startMs + trackBlockDays*dayMS
	tr := &positionTrack{
		midMs:  (startMs + endMs) / 2,
		halfMs: (endMs - startMs) / 2,
	}
	if body == RiseSetSun {
		tr.radiusAu = SunRadiusAU
	} else {
		tr.radiusAu = moonRadiusAU
	}

	for k := 0; k < trackNodes; k++ {
		x := math.Cos((jsnum.PI * float64(k)) / float64(trackNodes-1))
		tr.nodeX[k] = x
		end := 1.0
		if k == 0 || k == trackNodes-1 {
			end = 0.5
		}
		sign := 1.0
		if k%2 != 0 {
			sign = -1
		}
		tr.weight[k] = end * sign

		msf := tr.midMs + float64(tr.halfMs*x)
		ms := int64(msf)
		t := TTDaysSinceJ2000(ms) / 36525
		_, deps := Nutation(ctx, t)
		eps := (MeanObliquityArcsec(t) + deps) * ArcsecToRad

		var lonDeg, latDeg, distAu float64
		if body == RiseSetSun {
			p := GetSunPosition(ctx, ms)
			lonDeg, latDeg, distAu = p.Longitude, p.Latitude, p.Distance
		} else {
			p := GetMoonPositionForTrack(ctx, ms)
			lonDeg, latDeg, distAu = p.Longitude, p.Latitude, p.Distance/AuKm
		}
		lon := lonDeg * degToRad
		lat := latDeg * degToRad
		cosLat := distAu * math.Cos(lat)
		ex := cosLat * math.Cos(lon)
		ey := cosLat * math.Sin(lon)
		ez := distAu * math.Sin(lat)
		tr.px[k] = ex
		tr.py[k] = float64(math.Cos(eps)*ey) - float64(math.Sin(eps)*ez)
		tr.pz[k] = float64(math.Sin(eps)*ey) + float64(math.Cos(eps)*ez)
	}
	return tr
}

func (tr *positionTrack) position(ms float64) [3]float64 {
	x := (ms - tr.midMs) / tr.halfMs
	var nx, ny, nz, den float64
	for k := 0; k < trackNodes; k++ {
		dx := x - tr.nodeX[k]
		if dx == 0 {
			return [3]float64{tr.px[k], tr.py[k], tr.pz[k]}
		}
		q := tr.weight[k] / dx
		nx += float64(q * tr.px[k])
		ny += float64(q * tr.py[k])
		nz += float64(q * tr.pz[k])
		den += q
	}
	return [3]float64{nx / den, ny / den, nz / den}
}

type dayFrame struct {
	eqeq0      float64
	eqeq1      float64
	deltaTDays float64
	dayStartMs float64
}

func newDayFrame(ctx *EphemerisCtx, dayIndex int64) *dayFrame {
	dayStartMs := dayIndex * dayMS
	dayEndMs := dayStartMs + dayMS
	midMs := (dayStartMs + dayEndMs) / 2
	return &dayFrame{
		eqeq0:      equationOfEquinoxes(ctx, dayStartMs),
		eqeq1:      equationOfEquinoxes(ctx, dayEndMs),
		deltaTDays: TTDaysSinceJ2000(midMs) - float64(midMs-j2000NoonMS)/dayMS,
		dayStartMs: float64(dayStartMs),
	}
}

func (f *dayFrame) gast(ms float64) float64 {
	utDays := (ms - float64(j2000NoonMS)) / dayMS
	theta := float64(360 * jsnum.Mod(jsnum.Mod(0.7790572732640+float64(0.00273781191135448*utDays), 1)+jsnum.Mod(utDays, 1), 1))
	frac := (ms - f.dayStartMs) / dayMS
	eqeq := f.eqeq0 + float64((f.eqeq1-f.eqeq0)*frac)
	t := (utDays + f.deltaTDays) / 36525
	precession := 0.014506 + float64((4612.156534+float64((1.3915817+float64((-0.00000044+float64((-0.000029956+float64(-0.0000000368*t))*t))*t))*t))*t)
	gast := jsnum.Mod(theta+(eqeq+precession)/3600, 360)
	if gast < 0 {
		return gast + 360
	}
	return gast
}

func equationOfEquinoxes(ctx *EphemerisCtx, ms int64) float64 {
	t := TTDaysSinceJ2000(ms) / 36525
	dpsi, deps := Nutation(ctx, t)
	return dpsi * math.Cos((MeanObliquityArcsec(t)+deps)*ArcsecToRad)
}

const (
	maxTracks = 1024
	maxFrames = 8192
)

var (
	trackStore = store.New[string, *positionTrack](maxTracks, store.DefaultStripes, store.HashString)
	frameStore = store.New[int64, *dayFrame](maxFrames, store.DefaultStripes, store.HashInt64)
)

func trackFor(ctx *EphemerisCtx, body RiseSetBody, dayIndex int64) *positionTrack {
	blockIndex := floorDivInt(dayIndex, trackBlockDays)
	key := string(body) + "|" + jsnum.FormatInt(blockIndex)
	track, _ := trackStore.GetOrBuild(key, func() *positionTrack {
		return newPositionTrack(ctx, body, blockIndex)
	})
	return track
}

func frameFor(ctx *EphemerisCtx, dayIndex int64) *dayFrame {
	frame, _ := frameStore.GetOrBuild(dayIndex, func() *dayFrame {
		return newDayFrame(ctx, dayIndex)
	})
	return frame
}

func ClearRiseSetTracks() {
	trackStore.Clear()
	frameStore.Clear()
	scanCache.Clear()
	clearRiseSetEventCache()
}

type observerGeometry struct {
	sinPhi       float64
	cosPhi       float64
	equatorialAu float64
	polarAu      float64
	longitude    float64
}

func newObserverGeometry(location types.GeoLocation) observerGeometry {
	phi := location.Latitude * degToRad
	g := observerGeometry{
		sinPhi:    math.Sin(phi),
		cosPhi:    math.Cos(phi),
		longitude: location.Longitude,
	}
	flattened := float64(EarthFlatteningSquared * g.sinPhi * g.sinPhi)
	c := 1 / math.Sqrt(float64(g.cosPhi*g.cosPhi)+flattened)
	s := EarthFlatteningSquared * c
	heightKm := location.Elevation / 1000
	g.equatorialAu = (float64(EarthEquatorialRadiusKm*c) + heightKm) / AuKm * g.cosPhi
	g.polarAu = (float64(EarthEquatorialRadiusKm*s) + heightKm) / AuKm * g.sinPhi
	return g
}

func altitudeExcess(track *positionTrack, frame *dayFrame, geometry observerGeometry, ms float64) float64 {
	bodyVec := track.position(ms)
	local := (frame.gast(ms) + geometry.longitude) * degToRad
	cosLocal := Cos(local)
	sinLocal := Sin(local)

	x := bodyVec[0] - float64(geometry.equatorialAu*cosLocal)
	y := bodyVec[1] - float64(geometry.equatorialAu*sinLocal)
	z := bodyVec[2] - geometry.polarAu
	distance := math.Sqrt(float64(x*x) + float64(y*y) + float64(z*z))

	dot := (float64(float64(x*geometry.cosPhi)*cosLocal) +
		float64(float64(y*geometry.cosPhi)*sinLocal) +
		float64(z*geometry.sinPhi)) / distance
	clamped := dot
	if dot < -1 {
		clamped = -1
	} else if dot > 1 {
		clamped = 1
	}
	centre := float64(math.Asin(clamped) * radToDeg)
	return centre + float64((track.radiusAu/distance)*radToDeg) + RefractionNearHorizonDeg
}

func refine(f func(ms float64) float64, loMs, hiMs, fLoIn float64, wantRise bool) float64 {
	tolMs := rootToleranceDays * dayMS
	lo, hi, fLo := loMs, hiMs, fLoIn
	below := func(v float64) bool {
		if wantRise {
			return v < 0
		}
		return v >= 0
	}
	for hi-lo > tolMs {
		mid := (lo + hi) / 2
		fMid := f(mid)
		if below(fLo) == below(fMid) {
			lo, fLo = mid, fMid
		} else {
			hi = mid
		}
	}
	return (lo + hi) / 2
}

type dayEventPair struct {
	rise []float64
	set  []float64
}

func scanDay(ctx *EphemerisCtx, body RiseSetBody, location types.GeoLocation, dayIndex int64) dayEventPair {
	track := trackFor(ctx, body, dayIndex)
	frame := frameFor(ctx, dayIndex)
	geometry := newObserverGeometry(location)
	dayStart := float64(dayIndex * dayMS)
	dayEnd := dayStart + dayMS
	f := func(ms float64) float64 { return altitudeExcess(track, frame, geometry, ms) }

	var rise, set []float64

	var scan func(loMs, hiMs, fLo, fHi float64, depth int)
	scan = func(loMs, hiMs, fLo, fHi float64, depth int) {
		if fLo < 0 && fHi >= 0 {
			rise = append(rise, refine(f, loMs, hiMs, fLo, true))
			return
		}
		if fLo >= 0 && fHi < 0 {
			set = append(set, refine(f, loMs, hiMs, fLo, false))
			return
		}
		if depth >= 12 {
			return
		}

		span := float64(maxAltitudeSlopeDegPerDay * ((hiMs - loMs) / dayMS))
		highest := (fLo + fHi + span) / 2
		lowest := (fLo + fHi - span) / 2
		if fLo < 0 {
			if highest < 0 {
				return
			}
		} else if lowest >= 0 {
			return
		}

		midMs := (loMs + hiMs) / 2
		fMid := f(midMs)
		scan(loMs, midMs, fLo, fMid, depth+1)
		scan(midMs, hiMs, fMid, fHi, depth+1)
	}

	stepMs := scanStepDays * dayMS
	prevMs := dayStart
	prevF := f(prevMs)
	for ms := dayStart + stepMs; ; ms += stepMs {
		capped := math.Min(ms, dayEnd)
		value := f(capped)
		scan(prevMs, capped, prevF, value, 0)
		prevMs = capped
		prevF = value
		if capped >= dayEnd {
			break
		}
	}

	inside := func(list []float64) []float64 {
		out := make([]float64, 0, len(list))
		for _, e := range list {
			if e >= dayStart && e < dayEnd {
				out = append(out, e)
			}
		}
		sort.Float64s(out)
		return out
	}
	return dayEventPair{rise: inside(rise), set: inside(set)}
}

const maxScans = 20_000

var scanCache = store.New[string, dayEventPair](maxScans, store.DefaultStripes, store.HashString)

func DayEvents(ctx *EphemerisCtx, body RiseSetBody, direction int, location types.GeoLocation, dayIndex int64) []float64 {
	return append([]float64(nil), dayEventsShared(ctx, body, direction, location, dayIndex)...)
}

func dayEventsShared(ctx *EphemerisCtx, body RiseSetBody, direction int, location types.GeoLocation, dayIndex int64) []float64 {
	key := riseSetScanKey(body, location, dayIndex)
	pair, _ := scanCache.GetOrBuild(key, func() dayEventPair {
		return scanDay(ctx, body, location, dayIndex)
	})
	if direction == 1 {
		return pair.rise
	}
	return pair.set
}

func riseSetScanKey(body RiseSetBody, location types.GeoLocation, dayIndex int64) string {
	return string(body) + "|" +
		jsnum.FormatFloat(location.Latitude) + "|" +
		jsnum.FormatFloat(location.Longitude) + "|" +
		jsnum.FormatFloat(location.Elevation) + "|" +
		jsnum.FormatInt(dayIndex)
}

func floorDivInt(a, b int64) int64 {
	q := a / b
	if a%b != 0 && (a < 0) != (b < 0) {
		q--
	}
	return q
}
